### ISSUE TITLE
third part of C-style for has messed up scoping

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15161]
Migrated from [rt.perl.org#127463](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=127463) (status was 'open')

Searchable as RT127463$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



$ perl \-we 'for \(my $i = 1; $i \< 10; $i \+= my $k\) \{ $k = $i; print "$i\\n"; \}'
1
2
4
8

I'm not sure what's going on here\, but 'my $k' should probably not do that \(it
should always evaluate to undef\)\. In fact\, I would have expected the third part
of a C\-style for to have its own little scope\, like

&nbsp;  for \(my $i = 1; $i \< 10; do \{ $i \+= my $k \}\)

But for some reason $k leaks through to the loop body\, where it can be set
before officially being created by 'my'\. This is probably a variant of the 'my
$x if 0' problem\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.22.1:

Configured by mauke at Tue Dec 29 15:36:05 CET 2015.

Summary of my perl5 (revision 5 version 22 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=4.2.5-1-arch, archname=i686-linux
    uname='linux simplicio 4.2.5-1-arch #1 smp preempt tue oct 27 08:28:41 cet 2015 i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='5.3.0', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12, longdblkind=3
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/i686-pc-linux-gnu/5.3.0/include-fixed /usr/lib /lib
    libs=-lpthread -lnsl -lnm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -lnm -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.22.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.22'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.22.1:
    /home/mauke/usr/lib/perl5/site_perl/5.22.1/i686-linux
    /home/mauke/usr/lib/perl5/site_perl/5.22.1
    /home/mauke/usr/lib/perl5/5.22.1/i686-linux
    /home/mauke/usr/lib/perl5/5.22.1
    .


Environment for perl 5.22.1:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/mauke/perl5/perlbrew/bin:/home/mauke/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
    PERLBREW_BASHRC_VERSION=0.73
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_ROOT=/home/mauke/perl5/perlbrew
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @tux

On Thu\, 4 Feb 2016 17&#8203;:08&#8203;:28 \-0800\, "l\.mai@&#8203;web\.de \(via RT\)"
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> $ perl \-we 'for \(my $i = 1; $i \< 10; $i \+= my $k\) \{ $k = $i; print "$i\\n"; \}'
> 1
> 2
> 4
> 8
> 
> I'm not sure what's going on here\, but 'my $k' should probably not do that \(it
> should always evaluate to undef\)\. In fact\, I would have expected the third part
> of a C\-style for to have its own little scope\, like
> 
>   for \(my $i = 1; $i \< 10; do \{ $i \+= my $k \}\)
> 
> But for some reason $k leaks through to the loop body\, where it can be set
> before officially being created by 'my'\. This is probably a variant of the 'my
> $x if 0' problem\.

Normally\, the part inside the parens scopes to the block connected\, so
this does not surprise me\. Code like this I use on a very regular basis&#8203;:

$ cat t\.pl
use warnings;

if \(my @&#8203;foo = \(1\.\.4\)\) \{
&nbsp;    print "foo&#8203;: @&#8203;foo\\n";
&nbsp;    \}
print "$\_\\n" for @&#8203;foo;

$ perl t\.pl
Name "main&#8203;::foo" used only once&#8203;: possible typo at t\.pl line 6\.
foo&#8203;: 1 2 3 4

Which is completely in line with

$ cat t\.pl
use warnings;

for \(my $i = 1; $i \< 5; $i \+= my $k\) \{
&nbsp;    $k = $i;
&nbsp;    print "i&#8203;: $i\\n";
&nbsp;    \}
print "i&#8203;: $i\\n";
print "k&#8203;: $k\\n";

$ perl t\.pl
Name "main&#8203;::k" used only once&#8203;: possible typo at t\.pl line 8\.
Name "main&#8203;::i" used only once&#8203;: possible typo at t\.pl line 7\.
i&#8203;: 1
i&#8203;: 2
i&#8203;: 4
Use of uninitialized value $i in concatenation \(\.\) or string at t\.pl line 7\.
i&#8203;:
Use of uninitialized value $k in concatenation \(\.\) or string at t\.pl line 8\.
k&#8203;:

\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.23   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mauke

Am Do 04\. Feb 2016\, 23&#8203;:29&#8203;:24\, hmbrand schrieb&#8203;:
> 
> Normally\, the part inside the parens scopes to the block connected\, so
> this does not surprise me\. 

The difference is that

&nbsp;  for \(A; B; C\) \{ D \}

should be equivalent to

&nbsp;  A; while \(B\) \{ D \} continue \{ C \}

C is executed after A\, B\, and D\. That's why it's surprising that the loop body D can see and access variables declared in C\, even though their definitions haven't even been executed yet\.

The situation is analogous to

&nbsp;  % perl \-wE 'my $r; \{ say\(my $k\)\, last if $r; $k = "what"; $r = 1; redo \}'
&nbsp;  what

Here we also manage to make 'my $k' not evaluate to undef by arranging for the control flow not to match the lexical structure of the program\, executing 'my $k' after setting $k\.

This version is more obviously broken because the behavior of 'my \.\.\. if \.\.\.' is documented as undefined \("Here be dragons"\)\. But with C\-style 'for' you get this effect automatically and implicitly\, just by declaring variables in the third part\.

**Comment by p5pRT** [comment]
### From @ap

\* l\.mai@&#8203;web\.de via RT \<perlbug\-followup@&#8203;perl\.org> \[2016\-02\-05 12&#8203;:00\]&#8203;:
> Am Do 04\. Feb 2016\, 23&#8203;:29&#8203;:24\, hmbrand schrieb&#8203;:
> > Normally\, the part inside the parens scopes to the block connected\,
> > so this does not surprise me\.
>
> The difference is that
>
>   for \(A; B; C\) \{ D \}
>
> should be equivalent to
>
>   A; while \(B\) \{ D \} continue \{ C \}
>
> C is executed after A\, B\, and D\. That's why it's surprising that the
> loop body D can see and access variables declared in C\, even though
> their definitions haven't even been executed yet\.

Variables declared in the A part of for\(A;B;C\) do not remain declared
past the loop\. If you are serious about strict textual equivalence of
these constructs then that must also bother you\. Does it?

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @mauke

Am Fr 27\. Mai 2016\, 04&#8203;:19&#8203;:32\, aristotle schrieb&#8203;:
> \* l\.mai@&#8203;web\.de via RT \<perlbug\-followup@&#8203;perl\.org> \[2016\-02\-05 12&#8203;:00\]&#8203;:
> > Am Do 04\. Feb 2016\, 23&#8203;:29&#8203;:24\, hmbrand schrieb&#8203;:
> > > Normally\, the part inside the parens scopes to the block connected\,
> > > so this does not surprise me\.
> >
> > The difference is that
> >
> >   for \(A; B; C\) \{ D \}
> >
> > should be equivalent to
> >
> >   A; while \(B\) \{ D \} continue \{ C \}
> >
> > C is executed after A\, B\, and D\. That's why it's surprising that the
> > loop body D can see and access variables declared in C\, even though
> > their definitions haven't even been executed yet\.
> 
> Variables declared in the A part of for\(A;B;C\) do not remain declared
> past the loop\. If you are serious about strict textual equivalence of
> these constructs then that must also bother you\. Does it?

I'm not serious about strict equivalence\, let alone textual\.

What bothers me is the time traveling / backwards scoping of C\, which comes last in execution order\.

Anyway\, the equivalence wrt A can be easily restored by wrapping the whole thing in another scope&#8203;:

&nbsp;  do \{ A; while \(B\) \{ D \} continue \{ C \} \};

It's a bit harder to make 'my $x' not return undef in normal Perl code\.

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, May 30\, 2016 at 11&#8203;:55&#8203;:09AM \-0700\, l\.mai@&#8203;web\.de via RT wrote&#8203;:
> I'm not serious about strict equivalence\, let alone textual\.
> 
> What bothers me is the time traveling / backwards scoping of C\, which comes last in execution order\.
> 
> Anyway\, the equivalence wrt A can be easily restored by wrapping the whole thing in another scope&#8203;:
> 
>   do \{ A; while \(B\) \{ D \} continue \{ C \} \};
> 
> It's a bit harder to make 'my $x' not return undef in normal Perl code\.

Just as a data point\, the grammar for C\<for\(;;\)> is currently written
such that in terms of where \(lexically speaking\) lexical vars are
introduced and go out of scope\,

&nbsp;    for \(A; B; C\) \{ D \}

is equivalent to&#8203;:

&nbsp;    \{
&nbsp;        A;
&nbsp;        B;
&nbsp;        C;
&nbsp;        \{
&nbsp;            D;
&nbsp;        \}
&nbsp;    \}

\-\- 
Hofstadter's Law&#8203;: It always takes longer than you expect\, even when you
take into account Hofstadter's Law\.
